### ISSUE TITLE
Investigate our neighbors algorithm choice

### ISSUE THREAD

**Issue by ilan-gold** [scverse/scanpy#4131]
Right now, we are relying on the underlying implementation in `umap` for our neighbors calculation due to (potentially outdated) matintainability/usability concerns of almost every other library. We use PyNNDescent currently AFAICT:

https://github.com/scverse/scanpy/blob/bb4e605fd59263516f0c97d4bff7783f5fcfa541/src/scanpy/neighbors/__init__.py#L658-L665

in absence of a transformer but I think this should be reconsidered.

https://github.com/nmslib/hnswlib seems to be maintained if not amazingly documented
https://github.com/meilisearch/arroy lacks python bindings but looks to be maintained
https://github.com/facebookresearch/faiss seems to be maintained but like the first option lacks documentation it seems
https://github.com/zilliztech/pyglass top performer on https://ann-benchmarks.com/index.html#algorithms

Overall, we need to make a reasonable judgement about the speed-accuracy-maintainability tradeoff and try to figure out if we should make this change for our users.

See https://ann-benchmarks.com/index.html as well

**Comment by ilan-gold** [comment]
I think SoA (and from an scverse-adjacent person) seems to be https://arxiv.org/abs/2602.21247 assuming things pan out performance-wise the way the paper mentions. I've emailed the author about when the implementation will be available open-source, and it seems the answer is "soon, we would be excited to see this become a part of scanpy"

**Comment by iandriver** [comment]
I implemented the ppinn algorithm in the paper in rust and demoed it as another option for scanpy nn. [https://github.com/iandriver/pipnn-scanpy](https://github.com/iandriver/pipnn-scanpy) I'd be happy to keep working on it if interested, or wait for paper authors, but I was curious about the speed up and how it worked.

**Comment by ilan-gold** [comment]
Wow @iandriver - that's awesome!

I didn't expect PyNNDescent to have such a fast build time but the recall is moderately disturbing.  I reached out to the original author of the PiPNN paper and they said they may be open-sourcing their implementation soon. What is your assessment from what you've seen? Can PiPNN ever match the build time of PyNNDescent?

**Comment by iandriver** [comment]
I've spent a little more time optimizing and cross comparing, there is perhaps more to do including trying non Rust options (I found this as I'm looking for ways to use Rust to speed up [scanpy](https://github.com/iandriver/SingleRust/pull/1) so started with Rust). The [Readme](https://github.com/iandriver/pipnn-scanpy) has been updated. PiPNN as I've implemented/optimized it is now faster than PyNNDescent until ~800k cells but maintains perfect recall. If you tune pynndescent to have ~1 recall it is slower than PiPNN across the board. It seems better to me. If the Rust build makes sense I would like to as an option: 
`sc.pp.neighbors(adata, n_neighbors=15, use_rep="X_pca",
                transformer=PiPNNTransformer())`

<img width="532" height="286" alt="Image" src="https://github.com/user-attachments/assets/d3865680-d190-4043-838f-3d61d2e7f57f" />

**Comment by ilan-gold** [comment]
https://github.com/iandriver/pipnn-scanpy#full-sweep-5k800k

shows identical times to HNSW for build time however the paper claims 6-12x speedup (I think dependent on a "runs" hyperparameter, but the exact param is escaping me now).  What do you think is going on here?

**Comment by iandriver** [comment]
@ilan-gold  If you use the SIFT 1M dataset you see PiPNN is ~8x faster build time than FAISS (and pyglass). Pyglass is the fastest at recall. The [readme](https://github.com/iandriver/pipnn-scanpy) is updated with latest stats. Pyglass has been tricky as it should work on mac other chip sets but current wheel is AVX-512-only. Playing around with fixes for that on my fork currently.

**Comment by ilan-gold** [comment]
It seems like the takeaway (I think) is that PiPNN is always either the same or better speedwise at build time and has better accuracy at recall.  It seems, though, that it's speedup for build time, when present, is only for "high dimensional" data, which we generally don't have here on account of using PCs (i.e., <<128 dims).  Is that right?

**Comment by iandriver** [comment]
Yes that is right, my initial testing was focused on single cell data/scanpy nn testing ~50 PCA dims or <128 for sure. So the main outcome there was slight speed improvement while maintaining recall ~1. When you use the larger dataset (SIFT 1M) the build time improvement cited in the paper is there (6-12X). It is true at smaller scale but ends up ~1.5X. I have PiPNN vs FAISS vs Pyglass implementations that I can run now. If there is some dataset(s) that would be most relevant for a single cell nn benchmark let me know (I could sample across czi datasets of varying sizes or something like that).

**Comment by ilan-gold** [comment]
> When you use the larger dataset (SIFT 1M) the build time improvement cited in the paper is there (6-12X)

But even in your single-cell examples, you go up 800k.  But I guess the feature space is smaller i.e., PCs.  Maybe try it on 5 million or something?

The tahoe dataset would probably make sense: https://www.tahoebio.ai/news/open-sourcing-tahoe-100m - just using one plate or something?

**Comment by iandriver** [comment]
Ran all 4 (with caveats that I have made them all runnable on my Mac) using 1M-10M cells datasets from tahoe 100 dataset. Pipnn and pyglass are close except on peak memory use where pyglass wins by a little.  See [readme](https://github.com/iandriver/pipnn-scanpy/blob/main/bench/tahoe_dashboard.png). It seems clear to me that some version of pyglass and or Pipnn should be added to enable larger datasets to run faster with lower mem and the improvement  in recall performance relative to pynndescent. This has been a fun investigation, let me know if pursuing one or both is of interest and I can open a pull request to work on the details of implementation. 

<img width="401" height="271" alt="Image" src="https://github.com/user-attachments/assets/c73dca7e-70e0-4e47-a487-b34ddbf0b198" />

<img width="847" height="353" alt="Image" src="https://github.com/user-attachments/assets/ec5d30f6-74c2-4f76-a6c3-97585c69c717" />

**Comment by ilan-gold** [comment]
Hi @iandriver sorry for the delay here, this has been great. I mean to reply - I've been in touch with the PiPNN author and they will open-source their implementation soon including a GPU version.  I imagine we will use that as he reports they have more speedups on the way.  This has been invaluable work though to see the results reproduced and have a working implementation. I think I can fairly confidently say PiPNN will be our neighbors algorithm.  If you wanted to start giving people something to use, you can of course implement a the transformer API and post your package to PyPI :smile: - feel free to announce in zulip! See https://scanpy.readthedocs.io/en/1.11.x/api/generated/scanpy.pp.neighbors.html#scanpy.pp.neighbors's `transformer` argument

**Comment by ilan-gold** [comment]
P.S @iandriver do you want to jump on to our zulip? https://scverse.zulipchat.com/ It's a great platform and the rustar-aligner folks are there as well

**Comment by flying-sheep** [comment]
FYI: I chatted with a person named Luca on reddit who has some fundamental insight into the peculiarities of doing kNN -> tSNE/UMAP. Might be super helpful to hear them out. Small snippet (in case they want to reply themselves here, otherwise I’ll post what they told me on Reddit):

> The TLDR is that dimensionality reduction methods need, in several variants (UMAP, tSNE and others) the top-k neighbours of each entity. In several pipelines, this is done by running some vectorization of the original entity which might not actually cleanly vectorize, and then run an L2-based nearest neighbour that is then fed to a tSNE. For instance in sklearn implementation this is the default behaviour. In many datasets though, the vectorization -> L2 top k process is not the most appropriate, and it (1) adds the confounding aspect of how the vectorization is achieved and (2) is generally lossy. Often, there is already a definition of distance or similarity that can be directly applied to the entities, and build a nearest neighbour lookup based on that.

**Comment by LucaCappelletti94** [comment]
Hi @flying-sheep ! Happy to chat on the topic. The TLDR is: "use similarities and distances that apply best to your data, and this might not require any matrix densification. Any ANN that works well for that distance is a-okay then."

1) Data from instruments is, in a wide variety of cases, not matrix-shaped
2) Whether your `n`-samples dataset is `(n, h)`-matrix shaped or not it does not matter for several kinds of dimensionality reduction to get the `(n, d)`-embedding, as the only thing methods like tSNE or UMAP care about is an "affinity matrix", which in any usable implementation is of course not `(n, n)` but uses some top-k filter, so you get something that looks more like a `(n, k)` affinity graph, which is generally sparse and ragged. Other than efficiency, top-k filter is also used to trim and remove the heavy thin tail of the neighbours distribution (briefly, when an entry is similar to say, 60 entries and then has a brisk fall and all the other `n-60` entries have very small similarities, and while they have a very small value, they are MANY so they have weird effects on the resulting embedding)
3) When you have an `(n, h)` matrix, you can define the similarity or distance using euclidean or cosine, or generally speaking whatever you decide best applies to your feature vectors. The same applies to non-matrix-shaped data: you only need to define a consistent function `g` that will define how two entities in your dataset are similar or dissimilar. It does not even need to be metric or even symmetric, just consistent. For instance, per-human reviews of movies are not consistent, so you cannot get a good embedding out of them. Robust averages or medians in those cases are much better, when available.
4) If your data is not matrix shaped and you already have some good similarity/distance defined in the original space, embedding it into some `(n, h)` matrix and then running L2/cosine may not be the best choice to get the `top-k` neighbours from before
5) This is NOT an algorithmic choice (several ANN indices can be used, the most trivial of them being any LSH impl) but a distance/similarity choice

Bonus note: there is no particular reason to stop at the one affinity graph. Both t-SNE and UMAP can, in adequate implementations, handle any number of them. In particular regarding t-SNE (mostly because time in the day is limited and completing a there are already too many multi-tsne variants to evaluate without opening the door to other methods), I am carrying out a rather complete benchmark of current multi-affinity t-SNEs plus some ideas I have on the matter. If you have any set of samples for which you have multiple distinct measurements, you can create one visualization that includes all of them with the set of proper distances.

A reference I am benchmarking on this latest concept is: https://pmc.ncbi.nlm.nih.gov/articles/PMC8091681/

**Comment by iandriver** [comment]
Thanks @LucaCappelletti94 — useful framing, and it maps onto scanpy's setup pretty directly.

The split you're drawing between "which distance/similarity" and "which ANN index" is exactly what the `transformer=` API already separates: the transformer owns the metric and returns the sparse top-k graph, and the layout (UMAP/t-SNE) only ever sees that `(n, k)` graph — never an `(n, n)` matrix. So "use the similarity that fits your data" and "use whatever ANN is fast for that similarity" are independent knobs. On (2)/(5) especially: agreed, and nothing in the PiPNN work assumes the distance is metric or that the data had to be densified to get there.

Quick tooling update since @ilan-gold mentioned it: `pipnn` is now on PyPI (`pip install pipnn`, exposes `PiPNNTransformer`). I also put up `pyglass-portable` (`pip install pyglass-portable`, imports as `glass`) — a repackage of pyglass that builds and runs on Apple Silicon and non-AVX-512 x86, since the upstream wheel is AVX-512-only. Both drop straight into `sc.pp.neighbors(transformer=...)`.

On the multi-affinity bonus note — I read your reference (Do & Canzar / JVis). Two things that might be relevant to the benchmark you're running:

- The joint graph is assembled by looping over modalities and building a neighbor graph per modality, so neighbor construction scales with the number of modalities. That's precisely the ANN-swappable stage — umap-learn's `fuzzy_simplicial_set` takes precomputed neighbors, so any backend can feed it.
- On the t-SNE side, JTSNE builds those per-modality neighbors with sklearn's *exact* `NearestNeighbors` at k = 3·perplexity+1 (91 by default) — O(n²), and it dominates well before the layout does. Swapping in an ANN index there is a clean win at scale (in synthetic 50-d tests, several× by a few hundred thousand cells at matched recall). I reproduced the construction on a current stack since Jvis-learn no longer installs (pinned to sklearn ≤0.24.1) — happy to share the reimplementation and numbers if useful.
